src/Components/InputModel.re

```diff
@@ -28,13 +56,113 @@ let add = (~at as index, insert, text) => (
   index + String.length(insert),
 );
 
-let handleInput = (~text, ~cursorPosition) =>
-  fun
-  | "<LEFT>" => (text, max(0, cursorPosition - 1))
-  | "<RIGHT>" => (text, min(String.length(text), cursorPosition + 1))
-  | "<BS>" => removeBefore(cursorPosition, text)
-  | "<DEL>" => removeAfter(cursorPosition, text)
-  | "<HOME>" => (text, 0)
-  | "<END>" => (text, String.length(text))
-  | key when String.length(key) == 1 => add(~at=cursorPosition, key, text)
-  | _ => (text, cursorPosition);
+let removeCharBefore = (text, selection) => {
+  let (textSlice, _) = removeBefore(Selection.focus(selection), text);
+
+  (
+    textSlice,
+    Selection.collapseRelative(~text=textSlice, ~selection, Left(1)),
+  );
+};
+
+let removeSelection = (text, selection) => {
+  let (textSlice, focus) =
+    removeAfter(
+      Selection.rangeStart(selection),
+      text,
+      ~count=Selection.range(selection),
+    );
+
+  (textSlice, Selection.collapse(~text=textSlice, focus));
+};
+
+let removeCharAfter = (text, selection) => {
+  let (textSlice, focus) = removeAfter(Selection.focus(selection), text);
+
+  (textSlice, Selection.collapse(~text=textSlice, focus));
+};
+
+let collapseWord = (text, selection, direction) => {
+  let newSelection =
+    Selection.focus(selection)
+    |> direction(text)
+    |> Selection.collapse(~text);
+
+  (text, newSelection);
+};
+
+let extendWord = (text, selection, direction) => {
+  let newSelection =
+    Selection.focus(selection)
+    |> direction(text)
+    |> Selection.extend(~text, ~selection);
+
+  (text, newSelection);
+};
+
+let addCharacter = (key, text, selection) => {
+  let (newText, focus) = add(~at=Selection.focus(selection), key, text);
+
+  (newText, Selection.collapse(~text=newText, focus));
+};
+
+let replacesSelection = (key, text, selection) => {
+  let (textSlice, sliceSelection) = removeSelection(text, selection);
+  let (newText, focus) =
+    add(~at=Selection.focus(sliceSelection), key, textSlice);
+
+  (newText, Selection.collapse(~text=newText, focus));
+};
+
+let handleInput = (~text, ~selection, key) => {
+  switch (key, Selection.isCollapsed(selection)) {
+  | ("<LEFT>", true) => (
+      text,
+      Selection.collapseRelative(~text, ~selection, Left(1)),
+    )
```

Comment:
I really like how simple these all are. It makes it really easy to get a good overview of what it does for every given input. The function applications might be clearer in some cases though, but I've addressed that elsewhere

Reply:
:+1: